File: src/services/service_server.cpp
#include <string>

#include <ros/ros.h> // NodeHandle, ServiceServer

// custom service
#include <catkin_template/Service.h> // Service, ServiceRequest, ServiceResponse

// constants
static constexpr auto kNodeName = "cpp_service_server";

static constexpr auto kServiceName = "service";


bool handler(catkin_template::ServiceRequest& req, catkin_template::ServiceResponse& resp) {
  ROS_INFO("Service called on '%s' request = '%s'", ros::names::resolve(kServiceName).data(), req.request.data());
  resp.response = "cpp::service::response";
  return true;
}

int main(int argc, char **argv) {

  // initialize ROS Node
  ros::init(argc, argv, kNodeName);

  auto handle = ros::NodeHandle();

  // create and start service server
  ros::ServiceServer service = handle.advertiseService(kServiceName, handler);

  ROS_INFO("Service is ready on endpoint '%s'", ros::names::resolve(kServiceName).data());

  ros::spin();

  return 0;
}

File: scripts/services/service_server.py
#!/usr/bin/env python3
# encoding: utf-8

import sys
from typing import Final

import rospy

# custom service
from catkin_template.srv import Service, ServiceRequest, ServiceResponse

# constants
NODE_NAME: Final[str] = 'py_service_server'

SERVICE_NAME: Final[str] = 'service'


def handler(req: ServiceRequest) -> ServiceResponse:
  rospy.loginfo(f"Service called on '{rospy.resolve_name(SERVICE_NAME)}' request = '{req.request}'")
  return ServiceResponse(response='py::service::response')


def main() -> None:

  # initialize ROS Node
  rospy.init_node(NODE_NAME, sys.argv)

  # create service server
  rospy.Service(SERVICE_NAME, Service, handler)

  rospy.loginfo(f"Service is ready on endpoint '{rospy.resolve_name(SERVICE_NAME)}'")

  rospy.spin()


if __name__ == '__main__':
  main()
